akka-http-core/src/main/scala/akka/http/scaladsl/model/headers/HttpCookie.scala

```diff
@@ -113,10 +182,67 @@ final case class HttpCookie(
   /** Java API */
   def withHttpOnly(httpOnly: Boolean): headers.HttpCookie = copy(httpOnly = httpOnly)
   /** Java API */
+  def withSameSite(sameSite: jm.headers.SameSite): headers.HttpCookie = new HttpCookie(name, value, expires, maxAge, domain, path, secure, httpOnly, extension, Option(sameSite.asScala()))
+  /** Java API */
+  def withSameSite(sameSite: Optional[jm.headers.SameSite]): headers.HttpCookie = new HttpCookie(name, value, expires, maxAge, domain, path, secure, httpOnly, extension, sameSite.asScala.map(_.asScala()))
+  /** Java API */
   def withExtension(extension: String): headers.HttpCookie = copy(extension = Some(extension))
 }
 
 object HttpCookie {
+
+  def apply(
+    name:      String,
+    value:     String,
+    expires:   Option[DateTime] = None,
+    maxAge:    Option[Long]     = None,
+    domain:    Option[String]   = None,
+    path:      Option[String]   = None,
+    secure:    Boolean          = false,
+    httpOnly:  Boolean          = false,
+    extension: Option[String]   = None,
+    sameSite:  Option[SameSite] = None
+  ) = new HttpCookie(name, value, expires, maxAge, domain, path, secure, httpOnly, extension, sameSite)
+
+  @deprecated("for binary compatibility", since = "10.2.0")
+  def apply(
+    name:      String,
+    value:     String,
+    expires:   Option[DateTime],
+    maxAge:    Option[Long],
+    domain:    Option[String],
+    path:      Option[String],
+    secure:    Boolean,
+    httpOnly:  Boolean,
+    extension: Option[String]
+  ) = new HttpCookie(name, value, expires, maxAge, domain, path, secure, httpOnly, extension, None)
+
+  @deprecated("for binary compatibility", since = "10.2.0")
+  def unapply(cookie: HttpCookie) = Option((
+    cookie.name(),
+    cookie.value(),
+    cookie.expires,
+    cookie.maxAge,
+    cookie.domain,
+    cookie.path,
+    cookie.secure(),
+    cookie.httpOnly(),
+    cookie.extension
+  ))
+
+  def createFromPair(
+    pair:      HttpCookiePair,
+    expires:   Option[DateTime] = None,
+    maxAge:    Option[Long]     = None,
+    domain:    Option[String]   = None,
+    path:      Option[String]   = None,
+    secure:    Boolean          = false,
+    httpOnly:  Boolean          = false,
+    extension: Option[String]   = None,
+    sameSite:  Option[SameSite] = None): HttpCookie =
+    new HttpCookie(pair.name, pair.value, expires, maxAge, domain, path, secure, httpOnly, extension, sameSite)
+
+  @deprecated("Use createFromPair instead", "10.2.0")
```

Comment:
(this could become `private[headers]` as well again)

Reply:
This was part of `HttpCookie` object public API. Not sure if users should be using it like that, but I would prefer to only have it deprecated.

Reply:
I think we have two options here:
 * remove the default parameter values here, add `private[headers`, and rename the new one to `fromPair` so it's also source compatible
 * keep this one just deprecated and create a completely different method that only takes a `HttpCookiePair` and suggest to use the `withXYZ` methods to change other fields

Reply:
> remove the default parameter values here, add private[headers, and rename the new one to fromPair so it's also source compatible

Maybe this is slightly different from what @raboof explained here: https://github.com/akka/akka-http/pull/2928#discussion_r380949103

(where source compatibility is not tolerated)

> keep this one just deprecated and create a completely different method that only takes a `HttpCookiePair` and suggest to use the `withXYZ` methods to change other fields

Yes, I think we can create `def apply(pair: HttpCookiePair)` and add the missing `with*`  methods. I'm just unsure how this would be a better API, though. As a user, this:

```scala
val cookie = HttpCookie(pair, domain = Some("akka.io"), secure = true, httpOnly = true)
```

Looks more convenient than:

```scala
val cookie = HttpCookie(pair).withDomain("akka.io").withSecure(true).withHttpOnly(true)
```

I consider the former easier to read as well, and it makes fewer allocations.

But, anyhow, _overall_ API consistent wins here, therefore, if Akka HTTP is adopting the later style, I agree it would be better to have the same for `HttpCookie`.

WDYT?